src/methods/environment.js

```diff
@@ -459,33 +484,412 @@ export function ResolveThisBinding(realm: Realm): NullValue | ObjectValue | Abst
 }
 
-export function BindingInitialization(realm: Realm, node: BabelNode, value: Value, environment: void | LexicalEnvironment) {
+export function BindingInitialization(realm: Realm, node: BabelNodeLVal, value: Value, strictCode: boolean, environment: void | LexicalEnvironment) {
   if (node.type === "ArrayPattern") { // ECMA262 13.3.3.5
     // 1. Let iterator be ? GetIterator(value).
+    let iterator = GetIterator(realm, value);
+
     // 2. Let iteratorRecord be Record {[[Iterator]]: iterator, [[Done]]: false}.
+    let iteratorRecord = {
+      $Iterator: iterator,
+      $Done: false,
+    };
+
     // 3. Let result be IteratorBindingInitialization for ArrayBindingPattern using iteratorRecord and environment as arguments.
+    try {
+      IteratorBindingInitialization(realm, node.elements, iteratorRecord, strictCode, environment);
+    } catch (error) {
+      // 4. If iteratorRecord.[[Done]] is false, return ? IteratorClose(iterator, result).
+      if (iteratorRecord.$Done === false && error instanceof AbruptCompletion) {
+        throw IteratorClose(realm, iterator, error);
+      }
+      throw error;
+    }
+
     // 4. If iteratorRecord.[[Done]] is false, return ? IteratorClose(iterator, result).
+    // TODO: Why does this work?
+    if (iteratorRecord.$Done === false) {
+      let completion = new AbruptCompletion(new UndefinedValue(realm));
+      let actualCompletion = IteratorClose(realm, iterator, completion);
+      if (actualCompletion !== completion) {
+        throw actualCompletion;
+      }
+    }
+
     // 5. Return result.
-    throw new Error("TODO: Patterns aren't supported yet");
+    return;
   } else if (node.type === "ObjectPattern") { // ECMA262 13.3.3.5
+    // BindingPattern : ObjectBindingPattern
+
     // 1. Perform ? RequireObjectCoercible(value).
     RequireObjectCoercible(realm, value);
 
     // 2. Return the result of performing BindingInitialization for ObjectBindingPattern using value and environment as arguments.
-    throw new Error("TODO: Patterns aren't supported yet");
+
+    for (let property of node.properties) {
+      let env = environment ? environment : realm.getRunningContext().lexicalEnvironment;
+
+      // 1. Let P be the result of evaluating PropertyName.
+      let P = EvalPropertyName(property, env, realm, strictCode);
+
+      // 2. ReturnIfAbrupt(P).
+
+      // 3. Return the result of performing KeyedBindingInitialization for BindingElement using value, environment, and P as arguments.
+      KeyedBindingInitialization(realm, property.value, value, strictCode, environment, P);
+    }
   } else if (node.type === "Identifier") { // ECMA262 12.1.5
     // 1. Let name be StringValue of Identifier.
     let name = ((node: any): BabelNodeIdentifier).name;
 
     // 2. Return ? InitializeBoundName(name, value, environment).
-    return InitializeBoundName(realm, name, value, environment);
+    InitializeBoundName(realm, name, value, environment);
+    return;
```

Comment:
`InitializeBoundName` returns undefined anyway? There may have been a reason in development, but there’s no obvious reason now. I’ll put it back the way it was.